### PR TITLE
CMSSW_14_2_X Tau POG - BoostedDeepTau

### PR DESCRIPTION
**Pull request by gparida**
#### Key Updates:

1. **Changes to `RecoTauTag/RecoTau`:**
   - As per Michal's suggestion, renamed the model to **BoostedDeepTauV2p0**. There are no other models with "version 0" in the current repository but to be safe added a condition based on `year_ == 20161718` (short for 2016, 2017, 2018) to select the correct feature scaling values for BoostedDeepTau.
   - No working points (WP) are stored. Only the raw scores are retained for the final NanoAOD output.

2. **Changes to `PhysicsTools/NanoAOD`:**
   - Encountered issues when running the NanoSequence with the default IDs listed under `nanoAOD_boostedTau_switch` (`["mvaIso", "mvaIsoNewDM", "mvaIsoDR0p3", "againstEle"]`). Errors [1] appeared during execution, **so I removed them and tested exclusively with BoostedDeepTau** - which then worked.
   - **Validation:** To ensure that the integration works as expected, I compared the BoostedDeepTau scores for 3,000 events from a Run2-UL 2016 post Radion signal file. The results from my private implementation and the current one showed no discrepancies in the scores.
   - **Run-3 Integration:** The testing performed was focused on a Run-2 UL file, I am uncertain whether this tagger is fully integrated into the Run-3 NanoSequence. Perhaps it would be best if the experts had a look and recommended its proper integration or Run-3.
   - **Testing Instructions:** To run the NanoSequence for testing, please use the following config [0].

3. **Note unrelated to BoostedDeepTau:** The `pt` threshold for `boostedTau` was set at `40 GeV` which seems a bit high. Perhaps it can be lowered to `18 GeV` (same as HPS taus)?. 

Thanks and regards,
Ganesh
[0]`/afs/cern.ch/work/g/gparida/public/DeepBoostedTauCMSSWIntegration/name_BoostedDeepTauv2p0/CMSSW_14_2_X_2024-09-11-2300/src/nanoConfigs/NANO_NANO.py`
[1]
```
---- Begin Fatal Exception 14-Sep-2024 13:29:57 CEST-----------------------
An exception of category 'NoProductResolverException' occurred while
[0] Processing Event run: 1 lumi: 1 event: 3 stream: 0
[1] Running path 'NANOAODSIMoutput_step'
[2] Prefetching for module NanoAODOutputModule/'NANOAODSIMoutput'
[3] Prefetching for module CandMCMatchTableProducer/'boostedTauMCTable'
[4] Prefetching for module PATObjectCrossLinker/'linkedObjects'
[5] Prefetching for module PATTauRefSelector/'finalBoostedTaus'
[6] Prefetching for module PATTauIDEmbedder/'slimmedTausBoostedNewID'
[7] Prefetching for module PATTauDiscriminantCutMultiplexer/'patTauDiscriminationByElectronRejectionMVA6BoostedForNano'
[8] Calling method for module PATTauDiscriminationAgainstElectronMVA6/'patTauDiscriminationByElectronRejectionMVA6RawBoostedForNano'
Exception Message:
Cannot find EventSetup module to produce data of type "GBRForest" in
record "GBRWrapperRcd" with product label "RecoTauTag_antiElectronMVA_NoEleMatch_woGwoGSF_BL".
Please add an ESSource or ESProducer to your job which can deliver this data.
```
```
---- Begin Fatal Exception 14-Sep-2024 13:35:29 CEST-----------------------
An exception of category 'NoProductResolverException' occurred while
[0] Processing Event run: 1 lumi: 1 event: 3 stream: 0
[1] Running path 'NANOAODSIMoutput_step'
[2] Prefetching for module NanoAODOutputModule/'NANOAODSIMoutput'
[3] Prefetching for module CandMCMatchTableProducer/'boostedTauMCTable'
[4] Prefetching for module PATObjectCrossLinker/'linkedObjects'
[5] Prefetching for module PATTauRefSelector/'finalBoostedTaus'
[6] Prefetching for module PATTauIDEmbedder/'slimmedTausBoostedNewID'
[7] Prefetching for module PATTauDiscriminantCutMultiplexer/'rerunDiscriminationByIsolationNewDMMVABoostedForNano'
[8] Calling method for module PATTauDiscriminationByMVAIsolationRun2/'rerunDiscriminationByIsolationNewDMMVArawBoostedForNano'
Exception Message:
Cannot find EventSetup module to produce data of type "GBRForest" in
record "GBRWrapperRcd" with product label "RecoTauTag_tauIdMVAIsoDBnewDMwLT".
Please add an ESSource or ESProducer to your job which can deliver this data.
```

### CODE REVIEW

**Comment by mbluj** [comment]
@gparida thank you for preparing this PR. I will try to review it asap (this week?). Please find below a few general comments, request
* Old tauIDs cannot be removed for now as they should serve as reference for some time, but we can decide to store only WPs w/o raw outputs to reduce impact on size of samples. I will check what is wrong with them - I suppose there is an issue with GlobalTag
* Taus entering BoostedTau table was preselected based on pt (>40), new-decay-mode-finding and loosest WPs of (three versions) of old VSJet discriminants. In your development you loose the 1st of the requirements and remove the last which for sure inflates number of stored boostedTaus. Could you check what is impact on this using ~1k of ttbar events (e.g. using RelVal samples), please? Regardless, is lowering pT of boosted taus to 18GeV really justified?
* To keep number of boostredTaus in the control, please define minimal cut on BoostedDeepTauVSJets discriminator to be applied in selection. Similarity to standard DeepTauVSJets VVVLoose it should remove 80-90% of j->tau fakes and have efficiency of ~98% for genuine tauh's. This cut should be part of preselection above.
* year in version name: until now it referred to year of sample used to train the model - I would like to keep this convention. Which samples was used in case of boostedDeepTau, all 2016-2018?

File: PhysicsTools/NanoAOD/python/boostedTaus_cff.py [hunk 1]
```diff
@@ -5,17 +5,18 @@
...
 ##################### Import reusable funtions and objects from std taus ########
 from PhysicsTools.NanoAOD.taus_cff import _tauIdWPMask, tausMCMatchLepTauForTable, tausMCMatchHadTauForTable,tauMCTable
-
 ##################### User floats producers, selectors ##########################
 
 
 finalBoostedTaus = cms.EDFilter("PATTauRefSelector",
     src = cms.InputTag("slimmedTausBoosted"),
-    cut = cms.string("pt > 40 && tauID('decayModeFindingNewDMs') && (tauID('byVVLooseIsolationMVArun2DBoldDMwLT') || tauID('byVVLooseIsolationMVArun2DBnewDMwLT'))")
+    #cut = cms.string("pt > 40 && tauID('decayModeFindingNewDMs') && (tauID('byVVLooseIsolationMVArun2DBoldDMwLT') || tauID('byVVLooseIsolationMVArun2DBnewDMwLT'))")
```
Comment (mbluj) [inline PhysicsTools/NanoAOD/python/boostedTaus_cff.py:13]: Loosening this preselection should be justified and impact on size studied.
Please define vvvloose WP on BoostedTauID which will remove 80-90% of j->tau fakes to be used here

File: PhysicsTools/NanoAOD/python/boostedTaus_cff.py [hunk 2]
```diff
@@ -59,16 +60,25 @@
        idAntiEle2018 = _tauIdWPMask("againstElectron%sMVA6", choices=("VLoose","Loose","Medium","Tight","VTight"), doc= "Anti-electron MVA discriminator V6 (2018)")
 )
 
+#DeepBoostedTau ID raw score branches 
+_deepTauVars2018v2p7 = cms.PSet(
```
Comment (mbluj) [inline PhysicsTools/NanoAOD/python/boostedTaus_cff.py:64]: Please  be consistent with naming things: propose use always something like boostedDeepTauv2p0

File: RecoTauTag/RecoTau/python/tauIdWPsDefs.py [hunk 3]
```diff
@@ -56,6 +56,11 @@
     }
 }
 
+WORKING_POINTS_20161718_v2p0 = {
```
Comment (mbluj) [inline RecoTauTag/RecoTau/python/tauIdWPsDefs.py:59]: Add boosted in name of WPs (even if now trivial)

File: RecoTauTag/RecoTau/python/tools/runTauIdMVA.py [hunk 4]
```diff
@@ -21,14 +22,14 @@ class TauIDEmbedder(object):
...
-        "deepTau2017v2", "deepTau2017v2p1", "deepTau2018v2p5", "deepTau2026v2p5"
+        "deepTau2017v2", "deepTau2017v2p1", "deepTau2018v2p5","boosteddeepTau2018v2p0","deepTau2026v2p5"
     ])
 
     def __init__(self, process, debug = False,
                  originalTauName = "slimmedTaus",
                  updatedTauName = "slimmedTausNewID",
                  postfix = "",
-                 toKeep =  ["deepTau2017v2p1", "deepTau2018v2p5", "deepTau2026v2p5"],
+                 toKeep =  ["deepTau2017v2p1", "deepTau2018v2p5", "deepTau2026v2p5","boosteddeepTau2018v2p0"],
```
Comment (mbluj) [inline RecoTauTag/RecoTau/python/tools/runTauIdMVA.py:32]: Do not add it to default `toKeep` to not mix boosted and standard IDs. You can on the other hand add a test *py which will run boostedDeepTau as part of unit tests.

**Review by mbluj — COMMENTED** [review]
mostly cosmetics

**Comment by mbluj** [comment]
One more general comment: is there any reason to not run boostedDeepTau at miniAOD level and store it in both mini and nonoAOD?

**Comment by gparida** [comment]
Hi Michal, Thank you. Some quick clarifications

- For the first two points: Yes indeed the old IDs need to be stored. I just mentioned, to let you know, since they were causing errors and I had to remove them for the test.  For the pt threshold. I do not have a strong opinion myself.  I mentioned 18GeV since that was the threshold for HPS taus. In the 2016 version of an analysis (B2G-17-006) using boostedTaus the threshold used was 20 - which is what I do too for my analysis. Maybe it has some impact on the signal efficiency but I don't have the numbers handy.


> Could you check what is impact on this using ~1k of ttbar events (e.g. using RelVal samples):

I can try this when I get a chance. But just to understand by "impact" do you mean the impact on file size? 
 
 
>Which samples was used in case of boostedDeepTau, all 2016-2018?

Due to a lack of statistics, we used odd-numbered events for 2016, 2017 and 2018.
 
 >One more general comment: is there any reason to not run boostedDeepTau at miniAOD level and store it in both mini and nonoAOD?
 
 It was easier to do this implementation at NanoAOD level (running on MiniAOD) following what was done for `deepTauv2p5`. I do not know how this is done for MiniAOD.

**Comment by mbluj** [comment]
> * For the first two points: Yes indeed the old IDs need to be stored. I just mentioned, to let you know, since they were causing errors and I had to remove them for the test.  For the pt threshold. I do not have a strong opinion myself.  I mentioned 18GeV since that was the threshold for HPS taus. In the 2016 version of an analysis (B2G-17-006) using boostedTaus the threshold used was 20 - which is what I do too for my analysis. Maybe it has some impact on the signal efficiency but I don't have the numbers handy.
> 
> > Could you check what is impact on this using ~1k of ttbar events (e.g. using RelVal samples):
> 
> I can try this when I get a chance. But just to understand by "impact" do you mean the impact on file size?

Yes, please check impact on size of both boostedTau collection and full nanoAOD.
> 
> > Which samples was used in case of boostedDeepTau, all 2016-2018?
> 
> Due to a lack of statistics, we used odd-numbered events for 2016, 2017 and 2018.
> 
> > One more general comment: is there any reason to not run boostedDeepTau at miniAOD level and store it in both mini and nonoAOD?
> 
> It was easier to do this implementation at NanoAOD level (running on MiniAOD) following what was done for `deepTauv2p5`. I do not know how this is done for MiniAOD.

OK, so we will try sort it out at TauPOG side.

**Comment by mbluj** [comment]
Problem with old tauIDs is cased by GlobalTag you tried to use - it is GT for 106X releases while you run with 142X release. For testing it is useful to use auto conditions, here (for 2016) one can use `auto:run2_mc`.

**Comment by mbluj** [comment]
This PR is superseded by #174 (revised and rebased to the current master) => closing it.

**Pull request closed without merging**